File: samplejavaapp/src/main/java/com/yoox/samplejavaapp/main/MainActivity.java
package com.yoox.samplejavaapp.main;

import android.os.Bundle;
import android.view.View;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.RecyclerView;
import com.yoox.net.FilterableRequest;
import com.yoox.net.models.outbound.SearchResultItem;
import com.yoox.samplejavaapp.R;
import com.yoox.samplejavaapp.common.SearchResultItemAdapter;
import com.yoox.samplejavaapp.detail.DetailActivity;

public class MainActivity extends AppCompatActivity implements SearchResultItemAdapter.OnItemSelectedListener {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        View loader = findViewById(R.id.loader);
        RecyclerView recyclerView = findViewById(R.id.recyclerView);
        View previousPage = findViewById(R.id.previous_page);
        View nextPage = findViewById(R.id.next_page);
        View filterColor = findViewById(R.id.filter_color);
        View filterDesigner = findViewById(R.id.filter_designer);
        View filterCategory = findViewById(R.id.filter_category);
        View filterPrices = findViewById(R.id.filter_prices);
        View reset = findViewById(R.id.reset);

        SearchResultItemAdapter adapter = new SearchResultItemAdapter(this);
        recyclerView.setAdapter(adapter);

        MainViewModel mainViewModel = ViewModelProviders.of(this).get(MainViewModel.class);
        mainViewModel.getState()
                .observe(this, newData -> {

                    if (newData.isLoading()) {
                        loader.setVisibility(View.VISIBLE);
                    } else {
                        loader.setVisibility(View.INVISIBLE);
                    }

                    FilterableRequest request = newData.getRequest();

                    previousPage.setEnabled(newData.isPreviousEnabled());
                    nextPage.setEnabled(newData.isNextEnabled());
                    filterColor.setEnabled(newData.isFilterColorEnabled());
                    filterDesigner.setEnabled(newData.isFilterDesignerEnabled());
                    filterCategory.setEnabled(newData.isFilterCategoryEnabled());
                    filterPrices.setEnabled(newData.isFilterPriceEnabled());
                    adapter.replaceItems(newData.getItems());

                    previousPage.setOnClickListener(v -> mainViewModel.previousPage(request, newData));
                    nextPage.setOnClickListener(v -> mainViewModel.nextPage(request, newData));
                    filterColor.setOnClickListener(v -> mainViewModel.colorFilter(request, newData));
                    filterDesigner.setOnClickListener(v -> mainViewModel.designerFilter(request, newData));
                    filterCategory.setOnClickListener(v -> mainViewModel.categoryFilter(request, newData));
                    filterPrices.setOnClickListener(v -> mainViewModel.priceFilter(request, newData));
                    reset.setOnClickListener(v -> mainViewModel.loadItems());
                });

        mainViewModel.getError().observe(this, error ->
                Toast.makeText(this, error, Toast.LENGTH_LONG).show());
        mainViewModel.loadItems();


    }

    @Override
    public void onItemSelected(SearchResultItem item) {
        DetailActivity.start(this, item.getId());
    }
}

File: samplekotlinapp/src/main/java/com/yoox/samplekotlinapp/main/MainActivity.kt
package com.yoox.samplekotlinapp.main

import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.yoox.net.models.outbound.SearchResultItem
import com.yoox.samplekotlinapp.R
import com.yoox.samplekotlinapp.common.SearchResultItemAdapter
import com.yoox.samplekotlinapp.detail.DetailActivity
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val adapter = SearchResultItemAdapter(this::onItemSelected)
        recyclerView.adapter = adapter

        val mainViewModel = ViewModelProviders.of(this).get(MainViewModel::class.java)
        mainViewModel.state
            .observe(this, Observer { newData ->

                return@Observer when (newData) {
                    MainState.Loading -> {
                        renderLoading()
                    }
                    is MainState.Result -> {
                        renderResult(newData, adapter, mainViewModel)
                    }
                }
            })

        mainViewModel.error.observe(this, Observer { error ->
            Toast.makeText(this, error, Toast.LENGTH_LONG).show()
        })
        mainViewModel.loadItems()
    }

    private fun renderLoading() {
        loader.visibility = View.VISIBLE
    }

    private fun renderResult(
        newData: MainState.Result,
        adapter: SearchResultItemAdapter,
        mainViewModel: MainViewModel
    ) {
        loader.visibility = View.INVISIBLE

        val request = newData.request

        previous_page.isEnabled = newData.isPreviousEnabled
        next_page.isEnabled = newData.isNextEnabled
        filter_color.isEnabled = newData.isFilterColorEnabled
        filter_designer.isEnabled = newData.isFilterDesignerEnabled
        filter_category.isEnabled = newData.isFilterCategoryEnabled
        filter_prices.isEnabled = newData.isFilterPriceEnabled
        adapter.replaceItems(newData.items)

        previous_page.setOnClickListener { mainViewModel.previousPage(request, newData) }
        next_page.setOnClickListener { mainViewModel.nextPage(request, newData) }
        filter_color.setOnClickListener { mainViewModel.colorFilter(request, newData) }
        filter_designer.setOnClickListener { mainViewModel.designerFilter(request, newData) }
        filter_category.setOnClickListener { mainViewModel.categoryFilter(request, newData) }
        filter_prices.setOnClickListener { mainViewModel.priceFilter(request, newData) }
        reset.setOnClickListener { mainViewModel.loadItems() }
    }

    private fun onItemSelected(item: SearchResultItem) {
        DetailActivity.start(this, item.id)
    }
}
